libbeat/beat/event.go

```diff
@@ -68,12 +75,50 @@ func (e *Event) PutValue(key string, v interface{}) (interface{}, error) {
 		default:
 			return nil, errNoTimestamp
 		}
+	} else if subKey, ok := metadataKey(key); ok {
+		if subKey == "" {
+			switch meta := v.(type) {
+			case common.MapStr:
+				e.Meta = meta
+			case map[string]interface{}:
+				e.Meta = meta
+			default:
+				return nil, errNoMapStr
+			}
+		} else if e.Meta == nil {
+			e.Meta = common.MapStr{}
+		}
+		return e.Meta.Put(subKey, v)
 	}
 
-	// TODO: add support to write into '@metadata'?
 	return e.Fields.Put(key, v)
 }
 
 func (e *Event) Delete(key string) error {
+	if subKey, ok := metadataKey(key); ok {
+		if subKey == "" {
+			e.Meta = nil
+			return nil
+		}
+		if e.Meta == nil {
+			return nil
+		}
+		return e.Meta.Delete(subKey)
+	}
 	return e.Fields.Delete(key)
 }
+
+func metadataKey(key string) (string, bool) {
+	if !strings.HasPrefix(key, "@metadata") {
```

Comment:
I am wondering if it would be safer to check for `"@metadata."`? What if an event includes the key `"@metadata2`? If I understand correctly so far, it means that `Meta` gets deleted.

Reply:
I added a test case to show that `Delete(@metadataSpecial)` doesn't cause `Meta` to be deleted.

The reason it checks for `@metadata` is to allow for operations like `Put(@metadata, map[string]interface{}{x: y})` and `Get(@metadata)`. When it finds that `@metadata` as the prefix it then checks to see if the next character is a `.` so in effect it is checking for `@metadata.`.